### PR TITLE
VIA support for Preonic Rev3

### PR DESCRIPTION
**Pull request by mixedfeelings**
## Description

Add VIA support for Preonic Rev3. Inspired by @emstaple's Planck Rev6 Via Port (https://github.com/qmk/qmk_firmware/pull/9591), this is a direct copy of the standard Preonic stock layout with the number of available layers updated to 7 to match Preonic defaults.

I know this PR is likely to remain in limbo like the Planck Rev6 (https://github.com/qmk/qmk_firmware/pull/9144) until the OLKB VID issue, and possible the VIA Device Version issues are solved but I wanted to get a working version out there in case people wanted to try it.

Here is a link to the accompanying VIA JSON. I won't PR to VIA until this is approved on the QMK side and the VID and possibly device version can be updated:
https://github.com/mixedfeelings/keyboards/blob/preonic-rev3/src/olkb/preonic/rev3/olkb-preonic-rev3.json

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/preonic/keymaps/via/keymap.c
Comment (drashna) [inline keyboards/preonic/keymaps/via/keymap.c:38]: Via only supports 4 layers by default.  While this can be increased, the board actually has to have enough eeprom to do so. 

The simplest solution here is to reduce the number of layers to 4, base, raise, lower, and adjust. 

Also there are the `KC_FN13` and `KC_FN23` keycodes that perform the tri-layer stuff.

Comment (mixedfeelings) [reply keyboards/preonic/keymaps/via/keymap.c:38]: Thanks @drashna! I have found that both the Planck and Preonic can both handle all 6 layers in VIA well, but there are some issues with the 6th layer keymap - via doesn't recognize key codes like DF yet... so it all comes out as hex codes. they still work, but its not easily human readable which seems the point of VIA.

Anyhow, I can reduce it to 4 layers if you all think that is better, but it hasn't been an issue.

I am not sure I understand about `KC_FN13` and `KC_FN23`?

Comment (drashna) [reply keyboards/preonic/keymaps/via/keymap.c:38]: I think reducing it 4 layers may be best, but I'd really say that @jackhumbert should have the final say on that. 

And the keycodes listed do the tri-layer stuff. Oh, sorry, they're `FN_MO13` and `FN_MO23`.
https://github.com/qmk/qmk_firmware/blob/135d28d2fb3234dab3ef5353b47ec5838534e39d/quantum/via.c#L163-L183

Comment (mixedfeelings) [reply keyboards/preonic/keymaps/via/keymap.c:27]: Sorry for the delay. I reduced to 4 layers as suggested, changed RAISE and LOWER to FN_MO13 and FN_MO23 respectively, and removed some the additional keycodes not understood by VIA. Thank you for the advice and suggestions!

I will implement some of these in my Planck PR when i get my replacement PCB from OLKB group buy.

All seem to be working on my end, so just waiting for approval from @jackhumbert and a good VID for OLKB :)

Comment (josuemldnd23) [reply keyboards/preonic/keymaps/via/keymap.c:27]: @mixedfeelings how do i go about testing this via support on my Preonic?

Comment (mixedfeelings) [reply keyboards/preonic/keymaps/via/keymap.c:27]: @josuemldnd23  are you able to clone my branch and compile the firmware? otherwise i can send you what i have compiled.

Comment (natedunn) [reply keyboards/preonic/keymaps/via/keymap.c:27]: @mixedfeelings would also love to test if I can. I'm new to gmk, so I am looking into compiling this, but if you have something already, I would take that too. ;)

File: keyboards/preonic/keymaps/via/config.h
Comment (fauxpark) [inline keyboards/preonic/keymaps/via/config.h:20]:
```suggestion
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/preonic/keymaps/via/keymap.c
Comment (mixedfeelings) [reply keyboards/preonic/keymaps/via/keymap.c:27]: @josuemldnd23 @natedunn here is the firmware i am using: 
https://github.com/mixedfeelings/VIA-layouts/blob/main/OLKB/Preonic/rev3/firmware/preonic_rev3_via.bin

along with this VIA JSON:
https://github.com/mixedfeelings/keyboards/blob/preonic-rev3/src/olkb/preonic/rev3/olkb-preonic-rev3.json

File: keyboards/preonic/keymaps/via/config.h
Comment (drashna) [inline keyboards/preonic/keymaps/via/config.h:1]: Could you add a GPL2+ compatible license header here? 

For instance: 
https://github.com/qmk/qmk_firmware/blob/28ff51175b6d45ce5a8dbfe7ed6e7a12df3bc8a8/quantum/template/base/keyboard.c#L1-L15

File: keyboards/preonic/keymaps/via/keymap.c
Comment (drashna) [inline keyboards/preonic/keymaps/via/keymap.c:140]: This block isn't needed, since VIA handles it already. 
```sugestion
```

**Review by drashna — CHANGES_REQUESTED** [review]

File: keyboards/preonic/keymaps/via/config.h
Comment (mixedfeelings) [reply keyboards/preonic/keymaps/via/config.h:1]: I definitely can, but this is basically a direct copy of the default folder so i am not sure who deserves credit - certainly not me! It has commits from both you and Jack that i noticed. All I did was remove the references to the startup songs for the removed keymaps (dvorak and colemak). Please advise on credit and I can add it!

File: keyboards/preonic/keymaps/via/keymap.c
Comment (mixedfeelings) [reply keyboards/preonic/keymaps/via/keymap.c]: Done! thanks!

**Review by fauxpark — APPROVED** [review]

**Review by fauxpark — COMMENTED** [review]
Code itself looks good, but the VID situation still needs resolution.

**Comment by mixedfeelings** [comment]
Thanks for the vid/pid update @jackhumbert !!

I updated the JSON to match: 
https://github.com/mixedfeelings/keyboards/blob/preonic-rev3/src/olkb/preonic/rev3/olkb-preonic-rev3.json

Do you think this PR is ready to go?

**Comment by jackhumbert** [comment]
Perfect! I think so - the linter gave an error that I'm not familiar with, and I'm not sure the CI is actually building things in its test, but from looking at the changes themselves, it all seems fine!

**Review by jackhumbert — APPROVED** [review]

**Comment by fauxpark** [comment]
The lint error is due to missing readmes in each revision folder. AFAIK this is not a hard requirement at the moment.

**Pull request merged**
